File: src/test/java/EntityTest.java
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.minare.core.models.annotations.entity.*;
import com.minare.core.models.Entity;
import com.minare.core.entity.EntityFactory;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonArray;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import io.vertx.ext.mongo.MongoClient;

import java.security.MessageDigest;
import java.util.*;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@ExtendWith(VertxExtension.class)
public class EntityTest {
    private Vertx vertx;

    @Mock
    private MongoClient mongoClient;

    @BeforeEach
    void setup(VertxTestContext testContext) {
        vertx = Vertx.vertx();
        MockitoAnnotations.openMocks(this);
        // Configure any needed mock behavior here
        testContext.completeNow();
    }

    private Region createTestFixture() {
        Region region = new Region();
        region._id = "507f1f77bcf86cd799439011";
        region.name = "Test Region Alpha";
        region.zones = new ArrayList<>();

        Zone zone1 = new Zone();
        zone1._id = "507f1f77bcf86cd799439012";
        zone1.name = "Combat Zone";
        zone1.region = region;
        zone1.units = new ArrayList<>();
        zone1.buildings = new ArrayList<>();

        Zone zone2 = new Zone();
        zone2._id = "507f1f77bcf86cd799439013";
        zone2.name = "Safe Zone";
        zone2.region = region;
        zone2.units = new ArrayList<>();
        zone2.buildings = new ArrayList<>();

        Building hq = new Building();
        hq._id = "507f1f77bcf86cd799439014";
        hq.name = "Headquarters";
        hq.zone = zone1;
        //hq.position = new MapVector2();
        //hq.position._id = "507f1f77bcf86cd799439018";
        //hq.position.x = 100;
        //hq.position.y = 100;
        //hq.position.owner = hq;
        hq.statuses = new HashSet<>(Arrays.asList("fortified", "powered"));

        Building barracks = new Building();
        barracks._id = "507f1f77bcf86cd799439015";
        barracks.name = "Barracks";
        barracks.zone = zone1;
        //barracks.position = new MapVector2();
        //barracks.position._id = "507f1f77bcf86cd799439019";
        //barracks.position.x = 150;
        //barracks.position.y = 100;
        //barracks.position.owner = barracks;
        barracks.statuses = new HashSet<>();  // empty set

        MapUnit soldier = new MapUnit();
        soldier._id = "507f1f77bcf86cd799439016";
        soldier.name = "Elite Soldier";
        soldier.zone = zone1;
        soldier.position = new MapVector2();
        soldier.position._id = "507f1f77bcf86cd799439020";
        soldier.position.x = 120;
        soldier.position.y = 120;
        soldier.position.owner = soldier;
        soldier.statuses = new HashSet<>(Arrays.asList("alert", "armed"));
        soldier.offense = new JsonObject()
                .put("burn", true)
                .put("AP", false)
                .put("melee", true)
                .put("damage", new JsonArray().add(5).add(10).add(15));

        MapUnit medic = new MapUnit();
        medic._id = "507f1f77bcf86cd799439017";
        medic.name = "Field Medic";
        medic.zone = zone2;  // in safe zone
        medic.position = new MapVector2();
        medic.position._id = "507f1f77bcf86cd799439021";
        medic.position.x = 200;
        medic.position.y = 200;
        medic.position.owner = medic;
        medic.statuses = new HashSet<>(Arrays.asList("healing"));
        medic.offense = new JsonObject()
                .put("burn", false)
                .put("AP", false)
                .put("melee", false)
                .put("damage", new JsonArray().add(1));

        // Wire up all the references
        zone1.buildings.addAll(Arrays.asList(hq, barracks));
        zone1.units.add(soldier);
        zone2.units.add(medic);
        region.zones.addAll(Arrays.asList(zone1, zone2));

        return region;
    }

    // The expected serialized form (would be generated from actual output)
    private JsonArray createTestJson() {
        JsonArray documents = new JsonArray();

        // Root Region
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439011")
                .put("type", "Region")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("name", "Test Region Alpha")
                        .put("zones", new JsonArray()
                                .add(new JsonObject()
                                        .put("$ref", "entity")
                                        .put("$id", "507f1f77bcf86cd799439012"))
                                .add(new JsonObject()
                                        .put("$ref", "entity")
                                        .put("$id", "507f1f77bcf86cd799439013")))));

        // Safe Zone
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439013")
                .put("type", "Zone")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("buildings", new JsonArray())
                        .put("name", "Safe Zone")
                        .put("region", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439011"))
                        .put("units", new JsonArray()
                                .add(new JsonObject()
                                        .put("$ref", "entity")
                                        .put("$id", "507f1f77bcf86cd799439017")))));

        // Field Medic
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439017")
                .put("type", "MapUnit")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("name", "Field Medic")
                        .put("offense", new JsonObject()
                                .put("burn", false)
                                .put("AP", false)
                                .put("melee", false)
                                .put("damage", new JsonArray().add(1)))
                        .put("position", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439021"))
                        .put("statuses", new JsonArray().add("healing"))
                        .put("zone", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439013"))));

        // Medic Position Vector
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439021")
                .put("type", "MapVector2")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("owner", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439017"))
                        .put("x", 200.0)
                        .put("y", 200.0)));

        // Combat Zone
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439012")
                .put("type", "Zone")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("buildings", new JsonArray()
                                .add(new JsonObject()
                                        .put("$ref", "entity")
                                        .put("$id", "507f1f77bcf86cd799439014"))
                                .add(new JsonObject()
                                        .put("$ref", "entity")
                                        .put("$id", "507f1f77bcf86cd799439015")))
                        .put("name", "Combat Zone")
                        .put("region", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439011"))
                        .put("units", new JsonArray()
                                .add(new JsonObject()
                                        .put("$ref", "entity")
                                        .put("$id", "507f1f77bcf86cd799439016")))));

        // Barracks
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439015")
                .put("type", "Building")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("name", "Barracks")
                        .put("position", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439019"))
                        .put("statuses", new JsonArray())
                        .put("zone", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439012"))));

        // Barracks Position Vector
        //documents.add(new JsonObject()
        //        .put("_id", "507f1f77bcf86cd799439019")
        //        .put("type", "MapVector2")
        //        .put("version", 1)
        //        .put("state", new JsonObject()
        //                .put("owner", new JsonObject()
        //                        .put("$ref", "entity")
        //                        .put("$id", "507f1f77bcf86cd799439015"))
        //                .put("x", 150.0)
        //                .put("y", 100.0)));

        // Headquarters
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439014")
                .put("type", "Building")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("name", "Headquarters")
                        .put("position", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439018"))
                        .put("statuses", new JsonArray()
                                .add("powered")
                                .add("fortified"))
                        .put("zone", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439012"))));

        // HQ Position Vector
        //documents.add(new JsonObject()
        //        .put("_id", "507f1f77bcf86cd799439018")
        //        .put("type", "MapVector2")
        //        .put("version", 1)
        //        .put("state", new JsonObject()
        //                .put("owner", new JsonObject()
        //                        .put("$ref", "entity")
        //                        .put("$id", "507f1f77bcf86cd799439014"))
        //                .put("x", 100.0)
        //                .put("y", 100.0)));

        // Elite Soldier
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439016")
                .put("type", "MapUnit")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("name", "Elite Soldier")
                        .put("offense", new JsonObject()
                                .put("burn", true)
                                .put("AP", false)
                                .put("melee", true)
                                .put("damage", new JsonArray().add(5).add(10).add(15)))
                        .put("position", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439020"))
                        .put("statuses", new JsonArray()
                                .add("alert")
                                .add("armed"))
                        .put("zone", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439012"))));

        // Soldier Position Vector
        documents.add(new JsonObject()
                .put("_id", "507f1f77bcf86cd799439020")
                .put("type", "MapVector2")
                .put("version", 1)
                .put("state", new JsonObject()
                        .put("owner", new JsonObject()
                                .put("$ref", "entity")
                                .put("$id", "507f1f77bcf86cd799439016"))
                        .put("x", 120.0)
                        .put("y", 120.0)));

        return documents;
    }

    public JsonArray createLookupGraphQuery() {
        return new JsonArray();
    }

    //public Graph<Entity, DefaultEdge> createAncestorGraph() {
    //    return new DefaultDirectedGraph<Entity, DefaultEdge>();
    //}

    public JsonArray createUpdateQuery() {
        return new JsonArray();
    }

    @Test
    void testSerialize_success(VertxTestContext testContext) {
        Region region = createTestFixture();
        JsonArray expected = createTestJson();

        // Keep hash printing for determinism check
        String expectedHash = hashJsonObject(expected);

        region.serialize().onComplete(ar -> {
            if (ar.succeeded()) {
                JsonArray serialized = ar.result();
                String actualHash = hashJsonObject(serialized);

                testContext.verify(() -> {
                    // Print hashes for debugging
                    System.out.println("Expected hash: " + expectedHash);
                    System.out.println("Actual hash: " + actualHash);

                    // AssertJ deep comparison
                    assertThat(serialized)
                            .usingRecursiveComparison()
                            .isEqualTo(expected);

                    // Also verify hash equality for determinism
                    assertThat(actualHash).isEqualTo(expectedHash);

                    testContext.completeNow();
                });
            } else {
                testContext.failNow(ar.cause());
            }
        });
    }

    @Test
    void testUpdate_success(VertxTestContext testContext) {

    }

    private String hashJsonObject(JsonArray json) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(json.encode().getBytes());
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

@JsonTypeName("Region")
@EntityType("Region")
class Region extends Entity {
    @State
    @JsonProperty("zones")
    public List<Zone> zones;

    @State
    @JsonProperty("name")
    public String name;
}

@JsonTypeName("Zone")
@EntityType("Zone")
class Zone extends Entity {
    @JsonProperty("name")
    @State
    public String name;

    @State
    @JsonProperty("region")
    @ParentReference
    public Region region;  // back-reference

    @JsonProperty("units")
    @State
    @MutateStrict
    @ChildReference
    public List<MapUnit> units;

    @JsonProperty("buildings")
    @State
    @ChildReference
    public List<Building> buildings;
}

@JsonTypeName("Building")
@EntityType("Building")
class Building extends Entity {
    @JsonProperty("name")
    @State
    public String name;

    @JsonProperty("zone")
    @State
    public Zone zone;  // back-reference

    @JsonProperty("position")
    @State
    @ChildReference
    public MapVector2 position;

    @JsonProperty("statuses")
    @State
    public Set<String> statuses;
}

@JsonTypeName("MapUnit")
@EntityType("MapUnit")
class MapUnit extends Entity {
    @JsonProperty("name")
    @State
    public String name;

    @JsonProperty("zone")
    @State
    public Zone zone;  // back-reference

    @JsonProperty("position")
    @State
    @ChildReference
    public MapVector2 position;

    @JsonProperty("statuses")
    @State
    public Set<String> statuses;

    @JsonProperty("offense")
    @State
    public JsonObject offense;  // contains burn, AP, melee booleans and int array
}

@JsonTypeName("MapVector2")
@EntityType("MapVector2")
class MapVector2 extends Entity {
    @JsonProperty("x")
    @State
    public double x;

    @JsonProperty("y")
    @State
    public double y;

    @JsonProperty("owner")
    @State
    @ParentReference
    public MapUnit owner;  // back-reference to either Building or MapUnit
}

class TestEntityFactory implements EntityFactory {
    private final Map<String, Class<?>> classes;

    public TestEntityFactory() {
        classes = new HashMap<>();
        // Register our base types
        classes.put("region", Region.class);
        classes.put("zone", Zone.class);
        classes.put("building", Building.class);
        classes.put("mapunit", MapUnit.class);
        classes.put("mapvector2", MapVector2.class);
    }

    @Override
    public Class<?> useClass(String type) {
        return classes.get(type.toLowerCase());
    }

    @Override
    public Entity getNew(String type) {
        switch (type.toLowerCase()) {
            case "region":
                return new Region();
            case "zone":
                return new Zone();
            case "building":
                return new Building();
            case "mapunit":
                return new MapUnit();
            case "mapvector2":
                return new MapVector2();
            default:
                return new Entity();
        }
    }
}

File: minare-framework/src/test/EntityTest.kt
/** As of Redis PubSub refactor, the Entity interface has changed completely
 These tests helped us develop the graphing stuff. They served their purpose. And now, like adult mayflies, they die.
 Another reminder that until you know what your unit SUTs are, your tests are ephemeral as hell. **/


/**import com.google.inject.AbstractModule
import com.google.inject.Guice
import com.google.inject.Injector
import com.google.inject.name.Names
import kotlin.com.minare.core.entity.models.Entity
import com.minare.core.entity.factories.EntityFactory
import com.minare.core.entity.ReflectionCache
import com.minare.core.entity.annotations.Parent
import com.minare.persistence.EntityStore
import com.minare.persistence.MongoEntityStore
import io.vertx.core.Vertx
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import io.vertx.ext.mongo.MongoClient
import io.vertx.junit5.VertxExtension
import io.vertx.junit5.VertxTestContext
import kotlinx.coroutines.runBlocking
import java.security.MessageDigest
import org.assertj.core.api.Assertions.assertThat
import org.jgrapht.graph.DefaultDirectedGraph
import org.jgrapht.graph.DefaultEdge
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito

@ExtendWith(VertxExtension::class)
class EntityTest {
    private lateinit var vertx: Vertx
    private lateinit var testInjector: Injector

    @Mock
    private lateinit var mongoClient: MongoClient

    @Mock
    private lateinit var entityFactory: EntityFactory

    private lateinit var reflectionCache: ReflectionCache
    private lateinit var mongoEntityStore: MongoEntityStore
    private val entityCollection = "entities"

    val testFixtures: EntityTestFixtures = EntityTestFixtures()

    @BeforeEach
    fun setup(testContext: VertxTestContext) {
        vertx = Vertx.vertx()
        MockitoAnnotations.openMocks(this)


        reflectionCache = ReflectionCache()


        setupEntityFactoryMocks()


        reflectionCache.registerEntityClasses(
            listOf(
                Region::class,
                Zone::class,
                Building::class,
                MapUnit::class,
                MapVector2::class
            )
        )


        testInjector = Guice.createInjector(object : AbstractModule() {
            override fun configure() {
                bind(ReflectionCache::class.java).toInstance(reflectionCache)
                bind(MongoClient::class.java).toInstance(mongoClient)
                bind(EntityFactory::class.java).toInstance(entityFactory)


                bind(EntityStore::class.java).to(MongoEntityStore::class.java)


                bind(String::class.java)
                    .annotatedWith(Names.named("entityCollection"))
                    .toInstance(entityCollection)
            }
        })


        mongoEntityStore = testInjector.getInstance(MongoEntityStore::class.java)

        testContext.completeNow()
    }

    /**
     * Helper method to set up entity factory behavior
     */
    private fun setupEntityFactoryMocks() {

        Mockito.`when`(entityFactory.useClass("MapVector2")).thenReturn(MapVector2::class.java)
        Mockito.`when`(entityFactory.useClass("MapUnit")).thenReturn(MapUnit::class.java)
        Mockito.`when`(entityFactory.useClass("Zone")).thenReturn(Zone::class.java)
        Mockito.`when`(entityFactory.useClass("Region")).thenReturn(Region::class.java)
        Mockito.`when`(entityFactory.useClass("Building")).thenReturn(Building::class.java)
    }

    /**
     * Helper method to inject dependencies into entity instances
     */
    @Test
    fun testSerialize_success(testContext: VertxTestContext) {

        val region = testFixtures.createTestFixture(testInjector)

        val expected = testFixtures.createTestJson()
        val expectedHash = hashJsonObject(expected)

        testContext.verify {
            runBlocking {
                val serialized = region.serialize()
                val actualHash = hashJsonObject(serialized)


                println("Expected hash: $expectedHash")
                println("Actual hash: $actualHash")

                // AssertJ deep comparison
                assertThat(serialized)
                    .usingRecursiveComparison()
                    .isEqualTo(expected)

                // Also verify hash equality for determinism
                assertThat(actualHash).isEqualTo(expectedHash)

                testContext.completeNow()
            }
        }
    }

    @Test
    fun testBuildAncestorTraversalQuery_success(testContext: VertxTestContext) = runBlocking {
        val entityId = "507f1f77bcf86cd799439020" // MapVector2 ID
        val parentRefPaths = listOf("state.parentEntity")
        val expectedQuery = testFixtures.createExpectedAncestorTraversalQuery(entityId, "entities", parentRefPaths)

        // When
        val actualQuery = mongoEntityStore.buildAncestorTraversalQuery(entityId)

        // Then
        assertThat(actualQuery)
            .usingRecursiveComparison()
            .isEqualTo(expectedQuery)

        testContext.completeNow()
    }

    @Test
    fun testTransformResultsToEntityGraph(testContext: VertxTestContext) {
        // Given
        val jsonArrayResults = testFixtures.createMockAggregationResults()

        // Configure mocks to return appropriate entities
        Mockito.`when`(entityFactory.useClass("MapVector2")).thenReturn(MapVector2::class.java)
        Mockito.`when`(entityFactory.useClass("MapUnit")).thenReturn(MapUnit::class.java)
        Mockito.`when`(entityFactory.useClass("Zone")).thenReturn(Zone::class.java)
        Mockito.`when`(entityFactory.useClass("Region")).thenReturn(Region::class.java)
        Mockito.`when`(entityFactory.useClass("Building")).thenReturn(Building::class.java)

        // Set up entityFactory.getNew to return properly initialized entities
        Mockito.`when`(entityFactory.getNew("MapVector2")).thenAnswer {
            val entity = MapVector2()
            testInjector.injectMembers(entity)
            entity
        }

        Mockito.`when`(entityFactory.getNew("MapUnit")).thenAnswer {
            val entity = MapUnit()
            testInjector.injectMembers(entity)
            entity
        }

        Mockito.`when`(entityFactory.getNew("Zone")).thenAnswer {
            val entity = Zone()
            testInjector.injectMembers(entity)
            entity
        }

        Mockito.`when`(entityFactory.getNew("Region")).thenAnswer {
            val entity = Region()
            testInjector.injectMembers(entity)
            entity
        }

        Mockito.`when`(entityFactory.getNew("Building")).thenAnswer {
            val entity = Building()
            testInjector.injectMembers(entity)
            entity
        }

        try {
            // When
            val graph = mongoEntityStore.transformResultsToEntityGraph(jsonArrayResults)

            // Then
            testContext.verify {
                // Debug: print entity count
                println("Graph vertices count: ${graph.vertexSet().size}")

                // Verify vertex count
                assertEquals(6, graph.vertexSet().size, "Graph should have 6 vertices")

                // Rest of the test...

                testContext.completeNow()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            testContext.failNow(e)
        }
    }

    @Test
    fun testFindEntitiesForVersionUpdate(testContext: VertxTestContext) {
        // Given
        // Create a simple graph structure with entities and their relationships
        val graph = DefaultDirectedGraph<Entity, DefaultEdge>(DefaultEdge::class.java)

        // Create test entities with proper dependency injection
        val vector = createTestEntity(MapVector2::class.java, "507f1f77bcf86cd799439020", "MapVector2") {}

        val soldier = createTestEntity(MapUnit::class.java, "507f1f77bcf86cd799439016", "MapUnit") {}

        val zone = createTestEntity(Zone::class.java, "507f1f77bcf86cd799439012", "Zone") {}

        val region = createTestEntity(Region::class.java, "507f1f77bcf86cd799439011", "Region") {}

        val building = createTestEntity(Building::class.java, "507f1f77bcf86cd799439015", "Building") {}

        val building2 = createTestEntity(Building::class.java, "507f1f77bcf86cd799439014", "Building") {}

        // IMPORTANT: Set up the actual parent references in the entity objects
        // These need to be set since shouldBubbleVersionToParent() checks these fields,
        // not just the graph structure
        vector.parentEntity = soldier
        soldier.zone = zone
        zone.region = region
        building.zone = zone
        building2.zone = zone

        // Add entities to the graph
        graph.addVertex(vector)
        graph.addVertex(soldier)
        graph.addVertex(zone)
        graph.addVertex(region)
        graph.addVertex(building)
        graph.addVertex(building2)

        // Add relationships (edges)
        graph.addEdge(vector, soldier) // Vector points to soldier (parentEntity)
        graph.addEdge(soldier, zone)   // Soldier points to zone
        graph.addEdge(zone, region)    // Zone points to region
        graph.addEdge(building, zone)  // Building points to zone (but not as parent reference)
        graph.addEdge(building2, zone) // Building2 points to zone (but not as parent reference)

        // When
        val idsToUpdate = vector.findEntitiesForVersionUpdate(graph)

        // Then
        testContext.verify {
            println("MapVector2 parent fields: ${reflectionCache.getFieldsWithAnnotation<Parent>(MapVector2::class).map { it.name }}")
            println("MapUnit parent fields: ${reflectionCache.getFieldsWithAnnotation<Parent>(MapUnit::class).map { it.name }}")
            println("Zone parent fields: ${reflectionCache.getFieldsWithAnnotation<Parent>(Zone::class).map { it.name }}")
            println("Building parent fields: ${reflectionCache.getFieldsWithAnnotation<Parent>(Building::class).map { it.name }}")

            // We expect only entities in the parent reference chain to be updated
            assertEquals(4, idsToUpdate.size, "Only 4 entities should need updates")

            // Verify each ID is in the set
            assertTrue(idsToUpdate.contains("507f1f77bcf86cd799439020"), "Vector ID should be in update set")
            assertTrue(idsToUpdate.contains("507f1f77bcf86cd799439016"), "Soldier ID should be in update set")
            assertTrue(idsToUpdate.contains("507f1f77bcf86cd799439012"), "Zone ID should be in update set")
            assertTrue(idsToUpdate.contains("507f1f77bcf86cd799439011"), "Region ID should be in update set")

            // Building should not be included as they don't use @ParentReference for zone
            assertFalse(idsToUpdate.contains("507f1f77bcf86cd799439015"), "Headquarters should not be in update set")
            assertFalse(idsToUpdate.contains("507f1f77bcf86cd799439014"), "Barracks should not be in update set")

            // Complete the test
            testContext.completeNow()
        }
    }

    @Test
    fun testMutatePruning(testContext: VertxTestContext) {
        // Arrange - create the entity using Guice
        val mapVector = testInjector.getInstance(MapVector2::class.java)
        mapVector._id = "test_vector_id"

        // Get test cases
        val testCases = testFixtures.createMapVector2MutationTestCases()

        testContext.verify {
            // Act & Assert - test each case
            for ((index, testCase) in testCases.withIndex()) {
                runBlocking {
                    val prunedDelta = mapVector.getMutateDelta(testCase.inputDelta)

                    // Assert pruned correctly
                    assertThat(prunedDelta)
                        .usingRecursiveComparison()
                        .isEqualTo(testCase.expectedPrunedDelta)

                    println("Test case $index passed!")
                }
            }

            testContext.completeNow()
        }
    }

    @Test
    fun testFilterDeltaByConsistencyLevel(testContext: VertxTestContext) {
        // Create test entities using Guice injection
        val mapVector = testInjector.getInstance(MapVector2::class.java).apply {
            _id = "test_vector_id"
            version = 5 // Set a specific version for testing
        }

        val mapUnit = testInjector.getInstance(MapUnit::class.java).apply {
            _id = "test_unit_id"
            version = 5 // Match version with mapVector for consistency
        }

        // Define test cases
        val testCases = testFixtures.createConsistencyLevelTestCases()

        testContext.verify {
            // Act & Assert - test each case
            for ((index, testCase) in testCases.withIndex()) {
                runBlocking {
                    // Get the right entity based on test case type
                    val entity = when (testCase.entityType) {
                        "MapVector2" -> mapVector
                        "MapUnit" -> mapUnit
                        else -> throw IllegalArgumentException("Unknown entity type")
                    }

                    // Execute the filter function
                    val filteredDelta = entity.filterDeltaByConsistencyLevel(
                        testCase.inputDelta,
                        testCase.requestedVersion
                    )

                    // Assert filtered correctly
                    assertThat(filteredDelta)
                        .usingRecursiveComparison()
                        .isEqualTo(testCase.expectedFilteredDelta)

                    println("Consistency level test case $index passed!")
                }
            }

            testContext.completeNow()
        }
    }

    private fun hashJsonObject(json: JsonArray): String {
        try {
            val digest = MessageDigest.getInstance("SHA-256")
            val hash = digest.digest(json.encode().toByteArray())
            val hexString = StringBuilder()
            for (b in hash) {
                val hex = Integer.toHexString(0xff and b.toInt())
                if (hex.length == 1) hexString.append('0')
                hexString.append(hex)
            }
            return hexString.toString()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    private fun <T : Entity> createTestEntity(entityClass: Class<T>, id: String, type: String, configure: (T) -> Unit): T {
        val entity = entityClass.newInstance()
        entity._id = id
        entity.type = type
        configure(entity)
        testInjector.injectMembers(entity)
        return entity
    }
}**/